### PR TITLE
Disable Contact Role Automation for specific Opportunities

### PR DESCRIPTION
**Pull request by RadSFDeveloper**
# Critical Changes

# Changes
* A new checkbox field, Disable Contact Role Automation, is available on the Opportunity object. This field can be used to disable several aspects of NPSP Opportunity Contact Role automation on selected Opportunity records. The primary use cases include data migration, custom code execution, or as part of a custom integration. It's unchecked by default. Note that the field doesn't affect Contact Role automation from other installed packages or your own customizations.

# Issues Closed

# New Metadata
- CustomField: Opportunity.DisableContactRoleAutomation__c

# Deleted Metadata

### CODE REVIEW

**Comment by RadSFDeveloper** [comment]
**lurch: attach W-029765

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000007fPY9QAM" target="_blank">W-029765</a>

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 1]
```diff
@@ -730,9 +741,12 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
         }
 
         for (Opportunity eachOppty : currentOpportunities) {
-            if (eachOppty.Primary_Contact__c != null && eachOppty.AccountId != null
+            if (eachOppty.Primary_Contact__c != null
```
Comment (force2b) [inline src/classes/OPP_OpportunityContactRoles_TDTM.cls]: Would this method be easier to read if the check for `eachOppty.DisableContactRoleAutomation__c` was after the for using a `continue'`

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 2]
```diff
@@ -762,7 +776,9 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
 
         for (Opportunity eachOppty : currentOpportunities) {
             if (eachOppty.AccountId != null
-                && OPP_AutomatedSoftCreditsService.isOrganizationalAccount(oppAccountDetails.get(eachOppty.AccountId).npe01__SYSTEMIsIndividual__c)) {
+                && OPP_AutomatedSoftCreditsService.isOrganizationalAccount(oppAccountDetails.get(eachOppty.AccountId).npe01__SYSTEMIsIndividual__c)
```
Comment (force2b) [inline src/classes/OPP_OpportunityContactRoles_TDTM.cls]: Same question here about checking for `eachOppty.DisableContactRoleAutomation__c` after the `for` and using `continue`.

File: src/classes/RD_RecurringDonations.cls [hunk 3]
```diff
@@ -637,43 +637,45 @@ public class RD_RecurringDonations {
         Map<String, Id> contactOppToContactRole = new Map<String, Id>();
             
         Opportunity[] oppsFromRecurringDonations = [
-            SELECT id, npe03__Recurring_Donation__c, npe03__Recurring_Donation__r.npe03__Contact__c 
+            SELECT Id, npe03__Recurring_Donation__c, npe03__Recurring_Donation__r.npe03__Contact__c
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/RD_RecurringDonations.cls]: Great catch @jjbennett!

File: src/classes/STG_PanelHealthCheck_CTRL.cls [hunk 4]
```diff
@@ -570,24 +570,36 @@ public without sharing class STG_PanelHealthCheck_CTRL extends STG_Panel {
...
-
-            if (cOpp > 0) {
-                createDR(label.healthLabelOCRCheck, statusError, 
-                  string.format(label.healthDetailsOCRDupPrimary, new string[]{string.valueOf(cOpp)}), label.healthSolutionOCRDupPrimary);      
+            List<AggregateResult> ocrAggregateResults = [
+                SELECT COUNT(Id), OpportunityId 
+                FROM OpportunityContactRole 
+                WHERE 
+                    IsPrimary = true 
+                    AND Opportunity.DisableContactRoleAutomation__c = false
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/STG_PanelHealthCheck_CTRL.cls]: I agree, especially if the fixup utility is going to make the corrections regardless of the `eachOppty.DisableContactRoleAutomation__c` setting

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 5]
```diff
@@ -296,97 +284,84 @@ private class HH_OppContactRoles_TEST {
...
-        Opportunity o1 = new Opportunity(
+        insert new Opportunity(
             Name = 'o1',
             StageName = UTIL_UnitTestData_TEST.getClosedWonStage(),
             CloseDate = System.today(),
-            Primary_Contact__c = (string)con1.id);
-        insert o1;
-        Opportunity o2 = new Opportunity(
+            Primary_Contact__c = (String)con1.id);
+
```
Comment (force2b) [inline src/classes/HH_OppContactRoles_TEST.cls]: Looks like a line of spaces is being inserted here.

File: src/classes/HH_OppContactRoles_TEST.cls [hunk 6]
```diff
@@ -296,97 +284,84 @@ private class HH_OppContactRoles_TEST {
...
             StageName = UTIL_UnitTestData_TEST.getClosedWonStage(),
             CloseDate = System.today(),
-            Primary_Contact__c = (string)con2.id);
-        insert o2;        
+            Primary_Contact__c = (String)con2.id);
         Test.stopTest();
         
-        system.assertequals(2,(integer)[select count() from Opportunity]);
-        system.assertequals(2,(integer)[select count() from OpportunityContactRole]);
+        System.assertEquals(2, (Integer)[SELECT count() FROM Opportunity]);
```
Comment (force2b) [inline src/classes/HH_OppContactRoles_TEST.cls]: These don't actually be cast as `Integer` to work

File: src/classes/LD_LeadConvertOverride_TEST.cls [hunk 7]
```diff
@@ -784,6 +784,57 @@ private class LD_LeadConvertOverride_TEST {
         system.assertEquals(convertedContacts[0].Id, listOpp[0].Primary_Contact__c);
     }
 
+    /*******************************************************************************************************
+    * @description Verifies primary OCR is not created when Opportunity "Disable Contact Role Automation" is checked.
+    */
+    public static void shouldNotCreatePrimaryOcrsWhenOppDisableContactRoleAutomationIsChecked() {
```
[the remark this answers is not in the document]
Comment (force2b) [reply src/classes/LD_LeadConvertOverride_TEST.cls]: That's a good point. Lead Conversion would create a new Opp where the DisableOCRAutomation field should default to false, so theoretically all of the existing automation should always run when creating a new Opportunity during lead conversion.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 8]
```diff
@@ -406,6 +413,10 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
                 Opportunity newOpp = listOpps[i];
                 Opportunity oldOpp = listOldOpps[i];
 
+                if (newOpp.DisableContactRoleAutomation__c == true) {
```
[the remark this answers is not in the document]
Comment (RadSFDeveloper) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: Actually no, Opportunities are traversed based on the input `listOpps` that contain all Opps. If lines 416-418 are commented out, two new tests in `OPP_OpportunityContactRoles_TDTM`  fail. This method can be refactored, but I wanted to be on the safe side and make the minimal changes.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls (hunk 1 above)
Comment (RadSFDeveloper) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: The check for `Disable Contact Role Automation` is another check as verification the Primary Contact is not null, so I would treat them the same. Please let me know if that is not the case.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls (hunk 2 above)
Comment (RadSFDeveloper) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: The check for `Disable Contact Role Automation` is another check as verification the AccountId is not null, so I would treat them the same. Please let me know if that is not the case.

File: src/classes/RD_RecurringDonations.cls (hunk 3 above)
[the remark this answers is not in the document]
Comment (RadSFDeveloper) [reply src/classes/RD_RecurringDonations.cls]: Code is used throughout the class (`if` block on line [597](https://github.com/SalesforceFoundation/Cumulus/blob/80a008ea278798e40f07b9955c30c57981833d84/src/classes/RD_RecurringDonations.cls#L597) and block starting on line [1077](https://github.com/SalesforceFoundation/Cumulus/blob/80a008ea278798e40f07b9955c30c57981833d84/src/classes/RD_RecurringDonations.cls#L1077)). Is this code not valid any more?

**Comment by RadSFDeveloper** [comment]
@force2b , @jjbennett : Code changes have been reverted back for RD, duplicate primary OCR removal and Health Check reporting of duplicate primary OCRs. Can you please review again?

@sam-knox : Could you please check `Disable Contact Role Automation` HelpText and Description are valid based on the latest requirements? Also, requirements are reflected in the PR description.

**Comment by sam-knox** [comment]
@RadSFDeveloper - Use this for the new help text and description text:

"When selected, this field disables several aspects of NPSP Opportunity Contact Role automation, potentially resulting in incorrect or duplicate data. Refer to the product documentation for more details."

**Comment by RadSFDeveloper** [comment]
> @RadSFDeveloper - Use this for the new help text and description text:
> 
> "When selected, this field disables several aspects of NPSP Opportunity Contact Role automation, potentially resulting in incorrect or duplicate data. Refer to the product documentation for more details."

It has been done. Thanks @sam-knox .

File: src/classes/HH_OppContactRoles_TDTM.cls [hunk 9]
```diff
@@ -65,28 +77,28 @@ public class HH_OppContactRoles_TDTM extends TDTM_Runnable {
                // UNDONE: waiting to add the actual custom setting to use.
         	if (strRefreshSetting == null || strRefreshSetting == 'none')
         	   return null;
-        	boolean isOpenOnly = (strRefreshSetting == 'Open Opportunities');
+        	Boolean isOpenOnly = (strRefreshSetting == 'Open Opportunities');
```
Comment (jjbennett) [inline src/classes/HH_OppContactRoles_TDTM.cls]: Just a comment for when we refactor OCR's we should strip this code, I do not believe this feature was fully implemented.

File: src/classes/HH_OppContactRoles_TDTM.cls [hunk 10]
```diff
@@ -143,7 +155,7 @@ public class HH_OppContactRoles_TDTM extends TDTM_Runnable {
             
             // query for OppConRoles where primary=true and OppID in arg set
             // query changed to also include Role = default role, because Contact Merge scenarios were clearing isPrimary (not by our code)!
-            map<String,Id> contactOppToContactRole = new map<String,Id>();
+            Map<String,Id> contactOppToContactRole = new Map<String,Id>();
```
Comment (jjbennett) [inline src/classes/HH_OppContactRoles_TDTM.cls]: Since we fixed the spacing on the other maps, just going to mention the space missing here but up to you if you want to fix it.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls (hunk 1 above)
Comment (jjbennett) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: I think it's ok to add it to the existing conditional statement and not having a separate check, I agree that it is definitely hard to follow along the large conditional but that can be addressed in a separate refactor.

File: src/classes/HH_OppContactRoles_TDTM.cls [hunk 11]
```diff
@@ -283,31 +295,37 @@ public class HH_OppContactRoles_TDTM extends TDTM_Runnable {
...
+    * @param hhIds The list of Household Account Ids
+    * @return void
     ********************************************************************************************************/
-    public static void recreateHouseholdSoftCredits(list<ID> listHHId) {
-    
-        // find all Opps for the Accounts
-        list<Opportunity> listOpp = [select Id, AccountId from Opportunity where AccountId in :listHHId];
+    public static void recreateHouseholdSoftCredits(List<Id> hhIds) {
+        List<Opportunity> opps = [
+            SELECT Id, AccountId, DisableContactRoleAutomation__c
```
Comment (jjbennett) [inline src/classes/HH_OppContactRoles_TDTM.cls]: I can't remember if we spoke about this one but it looks like we are querying for the field here but then not using it. Do we want to delete the existing ocrs when automation is disabled?

File: src/classes/HH_OppContactRoles_TDTM.cls [hunk 12]
```diff
@@ -55,8 +55,20 @@ public class HH_OppContactRoles_TDTM extends TDTM_Runnable {
         if (!currentHouseholdsSettings.npo02__Household_Contact_Roles_On__c)
             return null;
 
-        list<Opportunity> opportunities = newlist;
-        list<Opportunity> oldOpportunities = oldlist;
+        List<Opportunity> opportunities = new List<Opportunity>();
+        List<Opportunity> oldOpportunities = new List<Opportunity>();
+        for (Integer i = 0, size = newList.size(); i < size; i++) {
+            Opportunity opp = (Opportunity) newList[i];
```
Comment (jjbennett) [inline src/classes/HH_OppContactRoles_TDTM.cls]: Not sure if you want to change this here but should we grab the old contact by id vs index?

File: src/classes/HH_OppContactRoles_TDTM.cls (hunk 11 above)
Comment (jjbennett) [reply src/classes/HH_OppContactRoles_TDTM.cls]: To recap our discussion of our review of this code, it looks like this method is called only by the account merge and contact merge triggers. We have the option to:
1. Do not delete the existing ocrs if automation is disabled
2. Continue to delete the existing ocrs but ensure the household are regenerated in this scenario and create a unit test to verify
3. Intelligently check the ocrs for duplicate household roles and remove the duplicates only

File: src/classes/HH_OppContactRoles_TDTM.cls (hunk 9 above)
Comment (RadSFDeveloper) [reply src/classes/HH_OppContactRoles_TDTM.cls]: `isOpenOnly` is used on line 84. Also, there is a unit test (`updateOppHHAccountRefreshOpen()`](https://github.com/SalesforceFoundation/Cumulus/blob/master/src/classes/HH_OppContactRoles_TEST.cls#L386) testing `Open Opportunities`. If this functionality is not used, I think we should create another WI to address removal of it. Thoughts?

Comment (jjbennett) [reply src/classes/HH_OppContactRoles_TDTM.cls]: I have such a terrible memory, ... strRefreshSetting  is being checked for null or none and then the boolean is the result of comparing to open opps here on 80 but the variable does not vary, it is set on line 38 and then never touched again.

File: src/classes/HH_OppContactRoles_TDTM.cls (hunk 11 above)
Comment (RadSFDeveloper) [reply src/classes/HH_OppContactRoles_TDTM.cls]: Even though the field is not used in the current method, Opportunities are passed to the household TDTM `run()` which expects `DisableContactRoleAutomation__c` field.
```
HH_OppContactRoles_TDTM hhocr = new HH_OppContactRoles_TDTM();
DmlWrapper dmlWrapper = hhocr.run(opps, null, TDTM_Runnable.Action.AfterInsert, Schema.SObjectType.Opportunity);      
```

Comment (RadSFDeveloper) [reply src/classes/HH_OppContactRoles_TDTM.cls]: Thanks for raising question @jjbennett ! The expected behaviour has been confirmed to be the same regardless of the `Disable Contact Role Automation` since Contact/Account merging is different than the automation around creation.

Comment (jjbennett) [reply src/classes/HH_OppContactRoles_TDTM.cls]: Thank you for confirming, no other questions. This looks good!

File: src/classes/HH_OppContactRoles_TDTM.cls (hunk 12 above)
Comment (RadSFDeveloper) [reply src/classes/HH_OppContactRoles_TDTM.cls]: In this example it is fine since `newList` is not retrieved again in the method, and `newList` and `oldList` are in sync to start with.

Comment (jjbennett) [reply src/classes/HH_OppContactRoles_TDTM.cls]: Gotcha! I was thinking our preference was to always grab the old contact using the id vs relying on the order being the same. Thanks for clarifying.

File: src/classes/HH_OppContactRoles_TDTM.cls [hunk 13]
```diff
@@ -55,8 +55,20 @@ public class HH_OppContactRoles_TDTM extends TDTM_Runnable {
         if (!currentHouseholdsSettings.npo02__Household_Contact_Roles_On__c)
             return null;
 
-        list<Opportunity> opportunities = newlist;
```
[the remark this answers is not in the document]
Comment (RadSFDeveloper) [reply src/classes/HH_OppContactRoles_TDTM.cls]: Styling changes have been applied:
- Enclosing `if` and `else` blocks has been done throughout the class.
- Re-styling SOQLs.

File: src/classes/HH_OppContactRoles_TDTM.cls (hunk 9 above)
Comment (RadSFDeveloper) [reply src/classes/HH_OppContactRoles_TDTM.cls]: Created W-031780.

File: src/classes/OPP_OpportunityContactRoles_TEST.cls [hunk 14]
```diff
@@ -175,11 +185,17 @@ private class OPP_OpportunityContactRoles_TEST {
...
     * @description Tests contact role creation for 1:1 model, using contact attribution.
     */
-    static testMethod void oppRolesForOneToOneContact() {
+    @isTest
+    private static void oppRolesForOneToOneContact() {
         //skip the test if Advancement is installed
         if(ADV_PackageInfo_SVC.useAdv()) return;
 
-        npe01__Contacts_and_Orgs_Settings__c testSettings = UTIL_CustomSettingsFacade.getContactsSettingsForTests(new npe01__Contacts_and_Orgs_Settings__c (npe01__Account_Processor__c = CAO_Constants.ONE_TO_ONE_PROCESSOR,npe01__Enable_Opportunity_Contact_Role_Trigger__c = true, npe01__Opportunity_Contact_Role_Default_role__c = OPPORTUNITY_CONTACT_ROLE_DONOR));
+        npe01__Contacts_and_Orgs_Settings__c testSettings = UTIL_CustomSettingsFacade.getContactsSettingsForTests(
```
[the remark this answers is not in the document]
Comment (RadSFDeveloper) [reply src/classes/OPP_OpportunityContactRoles_TEST.cls]: Assignment to the unused var has been removed.

File: src/classes/HH_OppContactRoles_TEST.cls (hunk 5 above)
Comment (RadSFDeveloper) [reply src/classes/HH_OppContactRoles_TEST.cls]: It is not visible in VSCode. Regardless, I re-wrote the statement to use one `insert` instead and removed the line completely. :)

**Comment by RadSFDeveloper** [comment]
@rponti-sforg , the Disable Contact Role Automation PR is ready for testing.
